jina/executors/crafters/image/segmenter.py

```diff
@@ -0,0 +1,205 @@
+        center = _restore_channel_axis(np.asarray(_center), self.channel_axis)
+        return [
+            dict(offset=0, weight=1., blob=tl.astype('float32'), location=(top_tl, left_tl)),
+            dict(offset=0, weight=1., blob=tr.astype('float32'), location=(top_tr, left_tr)),
+            dict(offset=0, weight=1., blob=bl.astype('float32'), location=(top_bl, left_bl)),
+            dict(offset=0, weight=1., blob=br.astype('float32'), location=(top_br, left_br)),
+            dict(offset=0, weight=1., blob=center.astype('float32'), location=(top_center, left_center)),
+        ]
+
+
+class SlidingWindowImageCropper(BaseSegmenter):
+    """
+    :class:`SlidingWindowImageCropper` crops the image with a sliding window.
+    """
+
+    def __init__(self,
+                 target_size: int,
+                 strides: Tuple[int, int],
+                 padding: bool = False,
+                 channel_axis: int = -1,
+                 *args,
+                 **kwargs):
+        """
+
+        :param target_size: desired output size. If size is a sequence like (h, w), the output size will be matched to
+            this. If size is an int, the output will have the same height and width as the `target_size`.
+        :param strides: the strides between two neighboring sliding windows. `strides` is a sequence like (h, w), in
+            which denote the strides on the vertical and the horizontal axis.
+        :param padding: If False, only patches which are fully contained in the input image are included. If True,
+            all patches whose starting point is inside the input are included, and areas outside the input default to
+            zero. The `padding` argument has no effect on the size of each patch, it determines how many patches are
+            extracted. Default is False.
+        """
+        super().__init__(*args, **kwargs)
+        self.target_size = target_size
+        if len(strides) != 2:
+            raise ValueError(f'strides should be a tuple of two integers: {strides}')
+        self.stride_h, self.stride_w = strides
+        self.padding = padding
+        self.channel_axis = channel_axis
+
+    def _add_zero_padding(self, img: 'np.ndarray') -> 'np.ndarray':
+        h, w, c = img.shape
+        ext_h = self.target_size - h % self.stride_h
+        ext_w = self.target_size - w % self.stride_w
+        return np.pad(img,
+                      ((0, ext_h), (0, ext_w), (0, 0)),
+                      mode='constant',
+                      constant_values=0)
+
+    def _location_generator_without_padding(self, img: 'np.ndarray') -> Generator[Tuple[int, int], None, None]:
+        height, width, _ = img.shape
+        top, left = (0, 0)
+        if isinstance(self.target_size, Tuple):
+            w_size, h_size = self.target_size
+        else:
+            w_size = h_size = self.target_size
+
+        while top + h_size <= height:
```

Comment:
I wonder if there is a more `pythonic` way to construct these `generators`

Reply:
using Python here will be very slow. numpy can do this directly, here is a code snippet of my previous implementation on sliding windows

```python
    def _get_all_sliding_window(self, img: 'np.ndarray'):
        extend_height = self.window_size - (img.shape[0]) % self.stride_height
        extend_wide = self.window_size - (img.shape[1]) % self.stride_wide

        input = np.pad(img, ((0, extend_height),
                             (0, extend_wide),
                             (0, 0)),
                       mode='constant', constant_values=0)
        expanded_input = np.lib.stride_tricks.as_strided(
            input,
            shape=(
                1 + int((input.shape[0] - self.window_size) / self.stride_height),
                1 + int((input.shape[1] - self.window_size) / self.stride_wide),
                self.window_size,
                self.window_size,
                3
            ),
            strides=(
                input.strides[0] * self.stride_height,
                input.strides[1] * self.stride_wide,
                input.strides[0],
                input.strides[1],
                1
            ),
            writeable=False
        )
        center_point_list = [
            [self.window_size / 2 + x * self.stride_wide, self.window_size / 2 + y * self.stride_height]
            for x in range(expanded_input.shape[0])
            for y in range(expanded_input.shape[1])]

        expanded_input = expanded_input.reshape((-1, self.window_size, self.window_size, 3))
        return [np.array(Image.fromarray(img)) for img in expanded_input], center_point_list
```

Reply:
ok, thanks, will change! 

Should we use `center_points` as the `location` or the coordinates of the `(top, left)` coordinates?

Reply:
let's keep `top, left`. `center_points` can be derived from it anyway

Reply:
Done!